File: Day245/medianOfTwoSortedArrays.java
package Day245;

public class medianOfTwoSortedArrays {

    public static double findMedianSortedArrays(int[] nums1, int[] nums2) {
        int x = nums1.length;
        int y = nums2.length;
        if (x >  y) {
            return findMedianSortedArrays(nums2, nums1);
        }
        int low = 0, high = x;

        while (low <= high) {
            int px = high - ((high-low)>>1);
            int py = (x+y + 1) / 2 - px;
            int mx = (px == 0) ? Integer.MIN_VALUE:nums1[px-1];
            int my = (py == 0) ? Integer.MIN_VALUE:nums2[py-1];
            int minX = (px == x) ? Integer.MAX_VALUE: nums1[px];
            int minY = (py == y) ? Integer.MAX_VALUE: nums2[py];
            if (mx <= minY && my <= minX) {
                if ((x+y)%2 == 0) {
                    return ((double)(Math.max(mx, my)+Math.min(minX, minY)))/2.0;
                } else {
                     return (double)Math.max(mx, my);
                } 
            } else if (mx > minY) {
                high = px -1;
            } else {
                low = px+1;
            }
        }
        return -1;
    }
    public static void main(String[] args) {
        int arr1[] = {1,3};
        int arr2[] = {2,3,6,8,9};
        System.out.println(findMedianSortedArrays(arr1, arr2));
    }
}

File: Day245/medianOfTwoSortedArrays.go
package main

import (
	"fmt"
	"math"
)

func findMedianSortedArrays(nums1 []int, nums2 [] int) float64 {
	x := len(nums1)
	y := len(nums2)
	if x > y {
		return findMedianSortedArrays(nums2, nums1)
	}
	low, high  := 0, x
	for low <= high {
		pX := high - ((high-low)>>1)
		pY := (x+y+1) / 2 - pX
		maxX := math.MinInt
		maxY := math.MinInt
		minX := math.MaxInt
		minY := math.MaxInt
		if pX != 0 {
			maxX = nums1[pX-1] 
		}
		if pY != 0 {
			maxY = nums2[pY-1]
		}
		if pX != x {
			minX = nums1[pX]
		}
		if pY != y {
			minY = nums2[pY]
		}
		if maxX <= minY && maxY <= minX {
			if ((x+y)%2 == 0) {
				return math.Max(float64(maxX), float64(maxY)) + math.Min(float64(minX), float64(minY)) / 2.0
			} else {
				return math.Max(float64(maxX), float64(maxY))
			}
		} else if (maxX > minY) {
			high = pX - 1
		} else {
			low = pX + 1
		}
	}
	return -1.0;
}

func main() {
	arr1 := []int{1,2}
	arr2 := []int{3,4}
	fmt.Println(findMedianSortedArrays(arr1,  arr2))
}